karoo_gp/engine.py

```diff
@@ -67,7 +134,7 @@ def predict(self, trees, X, X_hash=None):
         self._type_check(trees)
 
         # Sort sample columns by terminal
-        variables = self.model.terminals_.variables.keys()
+        variables = [t.label for t in self.model.get_nodes(('terminal'))]
```

Comment:
I setup `get_nodes` to accept Iterables with strings, so lists or tuples - I was under the impression tuples are preferred, but I could imagine situations where you'd want to pass a list or set. Think I should enforce?

There was a test using lists. I added tests for tuple and set for now.

Reply:
The problem is that `('terminal')` is not a `tuple`, but a string.  Since `.get_nodes` expects an iterable, and strings are iterable, it will check all individual letters (and probably find no matches):
```pycon
>>> def foo(it):
...   for elem in it:
...     if elem in {'terminals', 'functions'}: print(f'{elem} matches')
... 
>>> foo(['terminals', 'bar'])  # works with lists
terminals matches
>>> foo(('terminals', 'bar'))  # works with tuples
terminals matches
>>> foo(['terminals'])  # works with 1-elem lists
terminals matches
>>> foo(('terminals'))  # this doesn't work (it's just a string)
>>> foo(('terminals',))  # this is a 1-elem tuple (notice the comma)
terminals matches
```
As for which one is preferred:
* `list`s are used for homogeneous sequences that might have more or less elements (e.g. a list of names as str)
* `tuple`s for (generally) non-homogeneous groups with a fixed number of elements (e.g. `(name, surname, dob)`)

In this case a `list` seems ok.

Reply:
Aha, I see! I forgot that `(terminals) != (terminals,)`. Will update everything to use `list`.